macros/core/src/program/mod.rs

```diff
@@ -12,19 +13,33 @@ use syn::{
     Type, TypePath, Visibility,
 };
 
+mod args;
+
 /// Static Span of Program `impl` block
 static mut PROGRAM_SPAN: Option<Span> = None;
 
-pub fn gprogram(program_impl_tokens: TokenStream2) -> TokenStream2 {
+pub fn gprogram(args: TokenStream2, program_impl_tokens: TokenStream2) -> TokenStream2 {
     let program_impl = parse_gprogram_impl(program_impl_tokens);
     ensure_single_gprogram(&program_impl);
-    gen_gprogram_impl(program_impl)
+    let args = parse_args(args);
+    gen_gprogram_impl(program_impl, args)
 }
 
 #[doc(hidden)]
-pub fn __gprogram_internal(program_impl_tokens: TokenStream2) -> TokenStream2 {
+pub fn __gprogram_internal(args: TokenStream2, program_impl_tokens: TokenStream2) -> TokenStream2 {
     let program_impl = parse_gprogram_impl(program_impl_tokens);
-    gen_gprogram_impl(program_impl)
+    let args = parse_args(args);
+    gen_gprogram_impl(program_impl, args)
+}
+
+fn parse_args(args: TokenStream2) -> ProgramArgs {
+    syn::parse2(args).unwrap_or_else(|err| {
+        abort!(
+            err.span(),
+            "`gprogram` attribute can only contain `handle_reply` and `handle_signal` parameters: {}",
```

Comment:
How do you know here what went wrong? To my POV, this error should be raise in place where it has appropriate context

Reply:
Takes span from the error